src/Shared/FileMatcher.cs

```diff
@@ -2275,6 +2296,53 @@ private static int SkipSlashes(string aString, int startingIndex)
             return index;
         }
 
+        /// <summary>
+        /// Returns true if drive enumerating wildcard patterns are detected using the directory and wildcard parts.
+        /// </summary>
+        private static bool IsDriveEnumeratingWildcardPattern(string directoryPart, string wildcardPart)
+        {
+            int directoryPartLength = directoryPart.Length;
+            int wildcardPartLength = wildcardPart.Length;
+            if (directoryPartLength >= 3 &&                                                                  // Handles <drive letter>:<slashes>** cases
+                wildcardPartLength >= 2 &&
+                IsDrivePatternWithoutSlash(directoryPart[0], directoryPart[1]))
+            {
+                for (int i = 2; i < directoryPartLength-1; i++)
+                {
+                    if (!FileUtilities.IsAnySlash(directoryPart[i]))
+                    {
+                        return false;
+                    }
+                }
+
+                return IsFullFileSystemScan(directoryPart[directoryPartLength - 1], wildcardPart[0], wildcardPart[1]);
+            }
+            else if (directoryPartLength == 1 &&                                                             // Handles \** case
+                     wildcardPartLength >= 2 &&
+                     IsFullFileSystemScan(directoryPart[0], wildcardPart[0], wildcardPart[1]))
+            {
+                return true;
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Returns true if given characters follow a drive pattern without the slash (ex: C:).
+        /// </summary>
+        private static bool IsDrivePatternWithoutSlash(char firstValue, char secondValue)
```

Comment:
Since you were asking about high-performance implementations here, there are a few options besides this. It's not obvious to me what would be best, so I'm not telling you to change, but they might be interesting:

1. Take `(string, startIndex)` instead of individual characters, and do lookups via offsets from the `startIndex`.
2. Take `ReadOnlyMemory<char>` instead of individual characters, and do `AsMemory().Slice()` at the caller to pass just what to inspect. This is basically identical to the above but a bit more strongly typed (and on Framework it's a bit slower I think because of the lack of fast-span).